src/reader/asynchronous.rs

```diff
@@ -0,0 +1,1036 @@
+//! A module to handle the `AsyncReader`
+
+use async_recursion::async_recursion;
+#[cfg(feature = "encoding")]
+use encoding_rs::Encoding;
+use std::future::Future;
+use std::io;
+use std::marker::Unpin;
+use std::path::Path;
+use std::pin::Pin;
+use std::str::from_utf8;
+use std::task::{Context, Poll};
+use tokio::fs::File;
+use tokio::io::{AsyncBufRead, AsyncBufReadExt, BufReader};
+
+use crate::errors::{Error, Result};
+use crate::events::{BytesDecl, BytesEnd, BytesStart, BytesText, Event};
+
+use super::{is_whitespace, Decode, Decoder, NamespaceBufferIndex, TagState};
+
+impl<B: AsyncBufRead> Decode for AsyncReader<B> {
+    #[cfg(feature = "encoding")]
+    fn read_encoding(&self) -> &'static Encoding {
+        self.encoding
+    }
+
+    #[cfg(feature = "encoding")]
+    fn read_is_encoding_set(&self) -> bool {
+        self.is_encoding_set
+    }
+
+    #[cfg(feature = "encoding")]
+    fn write_encoding(&mut self, val: &'static Encoding) {
+        self.encoding = val;
+    }
+
+    #[cfg(feature = "encoding")]
+    fn write_is_encoding_set(&mut self, val: bool) {
+        self.is_encoding_set = val;
+    }
+}
+
+/// A low level encoding-agnostic XML event reader.
+///
+/// Consumes a `BufRead` and streams XML `Event`s.
```

Comment:
Should  this be "Consumes a `AsyncBufRead`..."?

Reply:
Yes.